File: backend/src/main/java/com/nusmotion/backend/dto/TickerTape.java
package com.nusmotion.backend.dto;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.time.OffsetDateTime;

/**
 * Alert with geographic location from /TickerTapes.
 */
public record TickerTape(
        @JsonProperty("Accident_Latitude") Double accidentLatitude,
        @JsonProperty("Accident_Longitude") Double accidentLongitude,
        @JsonProperty("Affected_Service_Ids") String affectedServiceIds,
        @JsonProperty("ID") String id,
        @JsonProperty("Message") String message,
        @JsonProperty("Priority") String priority,
        @JsonProperty("Status") String status,
        @JsonProperty("Display_From") OffsetDateTime displayFrom,
        @JsonProperty("Display_To") OffsetDateTime displayTo,
        @JsonProperty("Created_On") OffsetDateTime createdOn,
        @JsonProperty("Created_By") String createdBy
) {}

File: frontend/lib/data/models/ticker_tape.dart
class TickerTape {
  final double? accidentLatitude;
  final double? accidentLongitude;
  final String affectedServiceIds;
  final String id;
  final String message;
  final String priority;
  final String status;

  /// Validity start time (from backend "displayFrom").
  final DateTime? displayFrom;

  /// Validity end time (from backend "displayTo").
  final DateTime? displayTo;

  /// When this ticker was created (from backend "createdOn").
  final DateTime? createdOn;

  /// Who created this ticker (from backend "createdBy").
  final String? createdBy;

  const TickerTape({
    this.accidentLatitude,
    this.accidentLongitude,
    required this.affectedServiceIds,
    required this.id,
    required this.message,
    required this.priority,
    required this.status,
    this.displayFrom,
    this.displayTo,
    this.createdOn,
    this.createdBy,
  });

  factory TickerTape.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(String? raw) {
      if (raw == null) return null;
      return DateTime.tryParse(raw);
    }

    return TickerTape(
      accidentLatitude: (json['Accident_Latitude'] as num?)?.toDouble(),
      accidentLongitude: (json['Accident_Longitude'] as num?)?.toDouble(),
      affectedServiceIds: json['Affected_Service_Ids'] as String? ?? '',
      id: json['ID'] as String? ?? '',
      message: json['Message'] as String? ?? '',
      priority: json['Priority'] as String? ?? '',
      status: json['Status'] as String? ?? '',
      displayFrom: parseDate(json['Display_From'] as String?),
      displayTo: parseDate(json['Display_To'] as String?),
      createdOn: parseDate(json['Created_On'] as String?),
      createdBy: json['Created_By'] as String?,
    );
  }
}
